In [ ]:
def check_continuous_nodes(file_path):
    try:
        with open(file_path, 'r') as file:
            # 读取所有行并提取每行的两个节点序号
            nodes = set()
            for line in file:
                line = line.strip()
                if line:  # 确保不处理空行
                    parts = line.split()
                    if len(parts) == 2:  # 确保每行有两个节点序号
                        try:
                            node1, node2 = int(parts[0]), int(parts[1])
                            nodes.add(node1)
                            nodes.add(node2)
                        except ValueError:
                            print(f"文件中包含无效的数字格式：{line}")
                            return False
                    else:
                        print(f"文件中某行格式不正确，应包含两个节点序号：{line}")
                        return False

        # 如果没有提取到任何节点序号，返回 False
        if not nodes:
            print("文件中没有有效的节点序号。")
            return False

        # 将所有节点序号排序
        sorted_nodes = sorted(nodes)

        # 检查是否是连续的自然数
        start = sorted_nodes[0]
        end = sorted_nodes[-1]
        expected_nodes = set(range(start, end + 1))

        # 检查是否有缺失的节点
        missing_nodes = expected_nodes - nodes
        if missing_nodes:
            print(f"缺失的节点序号：{sorted(missing_nodes)}")
            return False
        else:
            print("所有节点序号是连续的，没有缺失的节点。")
            return True

    except FileNotFoundError:
        print(f"文件 {file_path} 未找到。")
    except Exception as e:
        print(f"发生错误：{e}")

# 替换为你的文件路径
file_path = 'data.txt'
check_continuous_nodes(file_path)

In [4]:
def analyze_graph(file_path):
    """
    分析图的稠密程度等信息。
    :param file_path: 存储图的文本文件路径。
    """
    # 初始化变量
    num_edges = 0  # 边的数量
    nodes = set()  # 节点集合

    # 读取文件并统计信息
    with open(file_path, 'r') as file:
        for line in file:
            line = line.strip()
            if line:  # 跳过空行
                nodes_in_edge = line.split()
                if len(nodes_in_edge) != 2:
                    raise ValueError(f"Invalid edge format in line: {line}")
                node1, node2 = nodes_in_edge
                nodes.add(node1)
                nodes.add(node2)
                num_edges += 1

    num_nodes = len(nodes)  # 节点的数量
    max_possible_edges = num_nodes * (num_nodes - 1) // 2  # 无向图的最大边数

    # 计算稠密程度
    density = num_edges / max_possible_edges if max_possible_edges > 0 else 0

    # 输出结果
    print(f"Number of nodes: {num_nodes}")
    print(f"Number of edges: {num_edges}")
    print(f"Graph density: {density}")

    return num_nodes, num_edges, density


if __name__ == "__main__":
    file_path = "data.txt"  # 替换为你的文件路径
    analyze_graph(file_path)

Number of nodes: 11316811
Number of edges: 85331845
Graph density: 1.3325792428114106e-06


## data_large: 
Number of nodes: 11316811 
Number of edges: 85331845
Graph density: 1.3325792428114106e-06